### PR TITLE
clang-format whole project + new CI workflow for checking style

### PR DESCRIPTION
**Pull request by piponazo**
Over the last years I discussed or mention to different people the possibility/dream to apply the clang-format style over the whole project and enforce the strict usage of the style by adding some CI checks or by any other means. With this Draft I would like to propose you to finally do it. 

Thanks to [this](https://github.com/marketplace/actions/clang-format-github-action) Github action, the effort I had to do is minimal. We just need to add a new workflow which uses that action, and we need to indicate which folders and files to inspect. Then the action uses our `.clang-format` file to make sure all the files have the proper format. You can see the result here:
https://github.com/Exiv2/exiv2/actions/runs/1989475924

Over the last years, the `.clang-format` file evolved as different people proposed different customisations. In this branch you have the opportunity to propose & discuss further changes to the format style. Once we reach an agreement we should avoid to touch that style again as much as possible.

Any other suggestions regarding clang-format and its enforcement for future contributions is welcome.

## Possible future work
- Add git hook to automatically apply clang-format before commiting (or something similar)? + instructions about how to setup this.

### CODE REVIEW

File: include/exiv2/xmpsidecar.hpp
Comment (neheb) [inline include/exiv2/xmpsidecar.hpp:89]: These ending comments I find interesting. In other codebases that I've seen they just have the define name as a comment. I think google's cpplint also complains about it.

File: app/exiv2app.hpp
Comment (neheb) [inline app/exiv2app.hpp:66]: Feels like a lot of these can be converted to std::pair + structured bindings.

File: samples/easyaccess-test.cpp
Comment (neheb) [inline samples/easyaccess-test.cpp:17]: formatting here is nasty. Can be fixed by adding a comma at the end of the last member.

File: src/tiffcomposite_int.hpp
Comment (neheb) [inline src/tiffcomposite_int.hpp:1665]: This feels like it should be merged. I couldn't get it to compile afterwards though.

**Comment by clanmills** [comment]
I have always supported the concept of reformatting the code when we reach 1.0 and when the 0.27-maintenance branch is at end-of-life.

When we agreed this in May 2018, we did not expect 0.27-maintenance to be the shipping branch in 2022.

@alexvanderberkel has proposed a Team Meeting.  We should discuss to shipping 'main' and how to end-life 0.27-maintenance.

File: src/tiffcomposite_int.hpp
Comment (piponazo) [reply src/tiffcomposite_int.hpp:1665]: I do not think clang-format can do this automatically. Imagine the following case:

```cpp
namespace a
{
  namespace b
  {
  }
  namespace c
  {
  }
}
```
It would be wrong for clang-format to try to attempt such conversions.

Please take into account that I do not want to make manual modifications in this PR. The changes applied by clang-format are already too large to add more manual changes here.

File: include/exiv2/xmpsidecar.hpp
Comment (piponazo) [reply include/exiv2/xmpsidecar.hpp:89]: I am not sure what is your point here. clang-format is just reformatting here the spaces between the `#endif` and the comment.

File: app/exiv2app.hpp
Comment (piponazo) [reply app/exiv2app.hpp:66]: Please, let's focus in this PR in the clang-format style. I am sure you will find many opportunities here for refactoring but these should be addressed in other PRs :wink:

File: samples/easyaccess-test.cpp
Comment (piponazo) [reply samples/easyaccess-test.cpp:17]: I agree on this. I invested some time trying to find a way to split the arrays in an homogeneous way (keeping always the same indentation, separating each item in a different line, etc) but it looks difficult to achieve. I was specially playing with the file `src/fujimn_int.cpp`. 

I'll keep investigating how to make the formatting of arrays more "beautiful". Suggestions here are welcome!

**Comment by piponazo** [comment]
@clanmills in commit 575826d36a5cccd90646ca75d194913144250c55 I removed some sample applications which were not considered in our CMake code. It looks like you were the author of most of them. Do you think we should keep some of them? (if so, I will adapt the cmake code to compile them)

File: samples/easyaccess-test.cpp
Comment (piponazo) [reply samples/easyaccess-test.cpp:17]: In the last commit (ad7f1784a460439bd7616594154cfb2f96b8b94f) I wanted to show you what's the effect of adding a comma after the last item in the array.

In most of the cases, it makes the formatting better (see `fujiOffOn`, `fujiTone` and others) and it split each item in a different line. However, there are some cases in which clang-format decides to "pack" several items in the same line (see `fujiSharpness` or `fujiFlashMode`). That's something I would really like to change if it is possible. I would like to see a more consistent behavior for all the cases.

Comment (neheb) [reply samples/easyaccess-test.cpp:17]: You can work around it by adding // at the end of each entry.

The issue interestinly enough has to do with the max characters per line. If you remove that it won't pack them.

Comment (hassec) [reply samples/easyaccess-test.cpp:17]: I really don't like adding `//` at the end of the lines. It sets a bad example, as people will then just start using comments to hack around what `clang-format` does.

Comment (piponazo) [reply samples/easyaccess-test.cpp:17]: @hassec I just had to add the `//` in few cases. If none of the items have comments at their right side, then it is not needed to add more comments. I just had to do that trick in cases like the following one, where there was a unique comment in the last item of the array: https://github.com/Exiv2/exiv2/commit/5ef2f8bd1e52e01093e8bf9d3a73cb3eb288021f#diff-8e58de4a7a00f0a4e00754c14143edab88e415b0e92b8545f977c5b1fe478681R1407. 

I agree with you that we should not start adding many of such "tricks" and expect from contributors to know them or apply them. 

On the other hand, our codebase is full of such arrays with many items inside, and I would really like to have a consistent formatting for them.

Comment (hassec) [reply samples/easyaccess-test.cpp:17]: I'm not sure if I like setting `ColumnLimit : 0` as that disables e.g. the reformatting of functions etc. 

take this example:

```cpp
// if limit is 120
int test(uint32_t LongNameForParameter1, double LongNameForParameter2, const float &LongNameForParameter3);

// if limit is 80
int test(uint32_t LongNameForParameter1, double LongNameForParameter2,
         const float &LongNameForParameter3);
```

this won't get reformatted to a common form as `clang-format` will simply use the input provided spacing if you set `ColumnLimit` to `0`

**Comment by hassec** [comment]
Thanks for working on this @piponazo, I'd love to see this. 

Regarding the style to use, is there any downside of just picking one of the default styles? Would avoid even opening the pandora's box of discussing individual settings.

File: samples/easyaccess-test.cpp
Comment (piponazo) [reply samples/easyaccess-test.cpp:17]: Right, that was also a concern I had. I think that what we can see right now is Okayish because I firstly applied clang-format to the whole project with the ColumnLimit set to 120 and in the last commits (with `ColumnLimit=0`) it kept the formatting somehow. 

It would probably better to just keep the `.clang-format` with the original version. The only drawback is that some arrays are not "perfectly" formatted.

Keep your suggestions & thought flowing! Thanks for all the feedback ❤️

**Comment by piponazo** [comment]
> Regarding the style to use, is there any downside of just picking one of the default styles? Would avoid even opening the pandora's box of discussing individual settings.

That's a good way of thinking. I'll play with the different styles and check if some of them is similar to what we currently have. Any preferences?

**Comment by hassec** [comment]
I'd be happy with any of them :+1:

**Comment by piponazo** [comment]
> no strong preference, I'd be happy with any of them +1

I have been playing with the different styles, but no one is good enough for me without some customisation. At the end I ended up using the customised options we already had in the `.clang-format` file.  I left some notes explaining why I think it is important to keep the different block of settings.

**Comment by piponazo** [comment]
Suggestion for reviewing the changes: "Use the hide whitespace" option
image

File: .clang-format
Comment (hassec) [inline .clang-format:10]: I'd remove and leave it as default, which is `Auto` and let's `clang-format` detect the standard

File: .clang-format
Comment (hassec) [inline .clang-format:17]: this is the default value, thus let's remove this

File: .clang-format
Comment (hassec) [inline .clang-format:16]: This value is connected to the above Indent, per default it's `-1`. Do we really need this diversion from default?

File: .clang-format
Comment (hassec) [inline .clang-format:24]: IMHO this setup of `BreackBeforeBraces` adds a lot of extra vertical empty space that's not really useful. Can we stick to the default?

File: .clang-format
Comment (hassec) [inline .clang-format:39]: this is the default value for Google

File: .clang-format
Comment (hassec) [inline .clang-format:35]: is default

File: .clang-format
Comment (hassec) [inline .clang-format:37]: Those two, I think, are quite useful for the coverage stuff, good catch! :+1:

File: .clang-format
Comment (hassec) [inline .clang-format:21]: This one is default I think

**Review by hassec — CHANGES_REQUESTED** [review]
couple default values that I'd remove 

The comments about which style to pick is just to voice it. This is exactly the tricky bit of subjectivity that I think isn't worth it and why I would have just picked plain `Google` plus the ones that have coverage implications.   

Leaving it up to you what to do :wink:   I'm just happy that we will have a style :tada:

File: .clang-format
Comment (piponazo) [reply .clang-format:24]: You are right, this is where subjectivity starts. I have a slight preference for breaking these braces in most of cases, but I could adapt to the Google style over time 😇 .

I'll take another iteration with your proposed changes.

**Comment by piponazo** [comment]
@hassec I applied all your suggestions except for the removal of `AllowShortFunctionsOnASingleLine: None`. I have a strong preference for keeping that one. Unless most of you guys prefer to remove it, my vote is for keeping it :nerd_face:

**Review by neheb — DISMISSED** [review]

**Comment by neheb** [comment]
I agree with @hassec about keeping it close to the given style. The current extras make sense.

File: .clang-format
Comment (hassec) [reply .clang-format:4]: @piponazo `IndendtWidth` is still specified to be 4, do you want to keep it?

File: .clang-format
Comment (hassec) [inline .clang-format:8]: `Standard : Auto` is default so we can actually remove this line

File: .github/workflows/on_push_clang_format.yml
Comment (piponazo) [reply .github/workflows/on_push_clang_format.yml:8]: The truth is that I just used the first Github Action that I encountered. I'll try this one out :+1:

File: .clang-format
Comment (piponazo) [reply .clang-format:4]: Hum ... I do not think 4 it is the default, I tried to remove it and reapply clang-format and it looks like the default is 2.

Anyways, some explicit values would not hurt :)

Comment (hassec) [reply .clang-format:4]: I was confused because I suggested to keep the Google default which is 2 and you have a 👍 but didn't change it, that's why I commented.

I think 4 spaces is a lot for C++ but if you insist on that indent, I'd say we should def. go for 120 column limit and also modify the the negative indent of AccessModifierOffset to be -2

Comment (piponazo) [reply .clang-format:4]: ahhh I see your point now. Actually, lately I am also using `2` in other projects and I tend to prefer that indentation rather than `4`. 
Just to be sure, you are proposing to just delete the `IndentWidth` and keep `ColumnLimit:120' , right ?

Comment (hassec) [reply .clang-format:4]: Yep 👍

File: .github/workflows/on_push_clang_format.yml
Comment (piponazo) [reply .github/workflows/on_push_clang_format.yml]: I have been playing a bit around with `clang-format-lint-action` and it works quite well! It runs faster than the other one and as you said we can think about integrating an automatic action to automatically apply clang-format before merging a PR. 

However, I would wait a bit for taking the step of introducing automatic commits related to the clang-format ... it scares me a bit what could happen without having some weeks/months of experience with the new clang-format settings we are defining here. 

What I can definitely do is write some documentation about the implications of enforcing clang-format now in the project in the file `CONTRIBUTING.md`.

**Review by neheb — APPROVED** [review]

File: .github/workflows/on_push_clang_format.yml
Comment (hassec) [reply .github/workflows/on_push_clang_format.yml]: @piponazo You are probably right that it's better to gain some experience with the action before we let it automatically commit to master. 

How would you propose the workflow should look?       
Let's say a user opens a PR and the pipeline fails, what would you suggest to them they do?     
What if they don't have `clang-format`?     
Or like in my case my system comes with 
```sh
❯ clang-format --version
clang-format version 13.0.1
```
That means for me there will be cases where I won't be able to fix the pipeline without installing a new `clang-format` version or trying to figure out the correct diff to apply manually from the failing ci log.   

All of that sounds like a bit too much friction to me.   

What if additionally to the check that you implemented, we add one more action which is triggered by a specific comment. 
E.g. if a comment includes `ExivBot plz format` the action will be triggered and applies the correct formatting to the PR.

Comment (piponazo) [reply .github/workflows/on_push_clang_format.yml]: I was thinking about this yesterday when I went to bed (I think I have a problem 🛌 ).

If a contributor creates a PR and the CI pipeline fails (this might happen not only to new contributors but also to experienced ones) we will have a a CI job indicating exactly what is the problem. At that point we can point them to the `CONTRIBUTING.md` file which has some notes about applying clang-format before summiting code.

Regarding the clang-format versions. I tried yesterday and clang-format-12 & clang-format-13 are producing the same results with the `.clang-format` file we have here. Of course, if they do not have clang-format installed, they will need to go to the LLVM download page and download the binaries for their platform: https://releases.llvm.org/download.html. I agree that this might create some friction for people who never used clang-format before. But hey if they managed to install git, compilers, cmake, conan, etc ... they should manage to install one tool more 😋 .

The main problems I see with the automatic approach are the following:
- A contributor creates a PR and we will not have a clang-format check at that point (the automatic formatting will happen after the PR is approved and merged). Since we have no formatting checks when the PR is created, we will see PRs that might contain some crazy formatting , reviewers might start to discuss again about formatting style, or it is simply more difficult to review the code because we see different formatting styles in each PR. For me that is one of the most important point about brining a unified and enforced clang-format: to easy the coding reviews.
- Let's say that we can "live" with the previous point. After the PR is approved and we click the merge button, the automatic commit happens. At this point at least 2 things might go wrong:
  - The automatic commit produce some changes that break the compilation. This barely happens but the probability is not 0%. It happens twice to me while working on this branch: Once due to the order of headers, and other time with some problem with preprocessor code and comments. 
  - The 2 problem I see is that we might not like the automatic formatting applied to the code and we have no chance to adapt the code in the PR (the PR will be already merged!). By forcing to have the clang-format job passing before merging, we have the chance to still make some suggestions about formatting during the coding review. For example: adding a comma after the last element of an array to have a nicer style after clang-format is applied. 
- We can end up having too many commits just applying clang-format and when you try to use `git blame` you end up seeing that many areas of the code were finally touched by a "ROBOT". I think that this is bad and should be avoided as much as possible. We should also recommend contributors to apply clang-format before committing. Or if they forget, it is always possible to do an interactive rebase and squash some commits together.

Comment (hassec) [reply .github/workflows/on_push_clang_format.yml]: Behavior between versions can definitely change, we've experienced this at work.            
Thus, to be on the safe side one has to have the correct version, which is something I don't like.  I personally don't want to have to download a specific version of clang-format for exiv2 and then make my editor somehow pick that version over my default one. 

I think the contributing guidelines can still say to apply `clang-format` if it is available on a user's system. That get's a good starting point for a review  

Otherwise, I agree with some of your points about the automatic commit, but my last comment was about a manually triggered action on a PR. 

This could be applied before a review if the original formatting is terrible, or at the end before merging. 
But it would still enable a review of the formatting before merging and it would trigger the CI so we know if it compiles. 
And the problem of having these commits in the history, yes that is a bit annoying but in most cases we can probably just squash & merge and thus won't have those commits.  

A PR that is big enough to have many commits that shouldn't be squashed probably comes from a user who format and/or rebase themselves.

Comment (piponazo) [reply .github/workflows/on_push_clang_format.yml]: Right, I think the approach about having a comment to trigger the clang-format in the PR might be a good solution to achieve the best of both sides. 

I'll switch this PR back to draft state and play a bit with that 😉

**Review by hassec — APPROVED** [review]

**Pull request merged**
